templates/usr/local/sbin/simp_generate_types.epp

```diff
@@ -3,132 +3,259 @@
+options.path     = nil
+options.quiet    = false
+options.syslog   = false
+options.timeout  = 300
 
 OptionParser.new do |opts|
   opts.on('-p', '--path TGT_PATH') do |path|
     options.path = path
-
-    if path == 'all'
-      options.target_environments = environment_paths.compact
-    else
-      path << '/' if (File.directory?(path) and path[-1] != '/')
-      options.target_environments = Array(environment_paths.find do |env|
-        path.start_with?("#{env}/")
-      end).compact
-    end
-
-    # Convert each discovered environment into a hash noting the path and environment name
-    options.target_environments.map!{|e| { :path => e, :env_name => File.basename(e) }}
   end
-  opts.on('-l', '--logfile LOGFILE') do |file|
+  opts.on('-g', '--generate', 'Read TGT_PATH and process all items') do
+    options.generate = true
+  end
+  opts.on('-l', '--logfile LOGFILE', 'Output to this logfile') do |file|
     options.logfile = file
   end
-  opts.on('-d', '--delay SECONDS') do |delay|
+  opts.on('-d', '--delay SECONDS', 'Default: 10 - Wait a random number of seconds up to this limit before starting the reader') do |delay|
     options.delay = delay.to_i
   end
-  opts.on('-s', '--syslog') do
+  opts.on('-s', '--syslog', 'Write to syslog') do
     options.syslog = true
   end
-  opts.on('-q', '--quiet') do
+  opts.on('-t', '--timeout TIMEOUT',
+          'Seconds to wait for a current run to complete before force-unlocking and continuing.',
+          '  * Default: 300',
+          '  * Minimum: 5'
+         ) do |tmout|
+    options.timeout = [5, tmout.to_i].max
+  end
+  opts.on('-q', '--quiet', 'No terminal output') do
     options.quiet = true
   end
+  opts.on('-m', '--message MESSAGE', 'The message to be written to TGT_PATH from -p') do |msg|
+    options.message = msg.strip
+  end
   opts.on('-h', '--help', 'This help message') do
     puts opts
     exit
   end
 
   options.help = opts.help
-
 end.parse!
 
-fh = File.open(options.logfile, 'w') if options.logfile
+def lock(lockfile)
+  locked = false
 
-unless options.path
-  msg = "Error: You must pass a path"
+  # From the API documentation for File/flock as the safe method for write
+  # locking
+  fh = File.open(lockfile, File::RDWR|File::CREAT, 0640)
 
-  $stderr.puts msg
-  $stderr.puts options.help
+  locked = fh.flock(File::LOCK_NB|File::LOCK_EX)
+  if locked
+    fh.rewind
 
-  log.error(msg) if options.syslog
+    fh.puts(Process.pid)
 
-  fh.puts(msg) if fh
+    fh.flush
+    fh.truncate(fh.pos)
+    fh.close
+  end
 
-  exit 1
+  return locked
 end
 
-if options.target_environments.empty?
-  msg = "Error: Could not find environment for '#{options.path}'"
+def unlock(lockfile, target_pid=nil)
+  if File.exist?(lockfile)
+    pid = 0
 
-  $stderr.puts msg
+    fh = File.open(lockfile, 'r')
+    pid = fh.read.strip.to_i
 
-  log.error(msg) if options.syslog
+    # Don't kill yourself
+    if target_pid.nil? && (pid != Process.pid)
+      fh.flock(File::LOCK_EX)
```

Comment:
You don't have the lock.  Why are you trying again without the LOCK_NB?  This will hang

Reply:
Removing the locks in `unlock`. Re-reading the logic, we just need to spin when locking.